### PR TITLE
feat: support IP rate limiting

### LINKED ISSUE #58: [feature] support IP rate limiting rule
Like if there are more than X (like 100) requests/second, then block the IP for Y seconds (like 10 minutes)

### PR DESCRIPTION
**Pull request by love98ooo**
Fix: https://github.com/casbin/caswaf/issues/58

P.S. update antd to ~4.24.0

### CODE REVIEW

File: controllers/rule.go [hunk 1]
```diff
@@ -131,6 +131,8 @@ func checkExpressions(expressions []*object.Expression, ruleType string) error {
 		return checkWafRule(values)
 	case "IP":
 		return checkIpRule(values)
+	case "IpRate":
```
Comment (nomeguy) [inline controllers/rule.go]: Label is "IP Rate Limiting"

File: controllers/rule.go [hunk 2]
```diff
@@ -157,3 +159,19 @@ func checkIpRule(ipLists []string) error {
 	}
 	return nil
 }
+
+func checkIpRateRule(expressions []*object.Expression) error {
+	if len(expressions) != 1 {
```
Comment (nomeguy) [inline controllers/rule.go:189]: Will this be zero?

File: controllers/rule.go [hunk 3]
```diff
@@ -157,3 +159,19 @@ func checkIpRule(ipLists []string) error {
...
+		return errors.New("IpRate rule should have only one expression")
```
Comment (nomeguy) [inline controllers/rule.go]: Better to limit this in frontend too

File: web/src/RuleEditPage.js [hunk 4]
```diff
@@ -86,7 +96,7 @@ class RuleEditPage extends React.Component {
                   {value: "WAF", text: "WAF"},
                   {value: "IP", text: "IP"},
                   {value: "User-Agent", text: "User-Agent"},
-                  // {value: "frequency", text: "Frequency"},
+                  {value: "IpRate", text: "IP Rate"},
```
Comment (nomeguy) [inline web/src/RuleEditPage.js]: Label is "IP Rate Limiting"

File: web/src/RuleEditPage.js [hunk 5]
```diff
@@ -131,6 +141,17 @@ class RuleEditPage extends React.Component {
                 />
               ) : null
             }
+            {
+              this.state.rule.type === "IpRate" ? (
+                <IpRateRuleTable
+                  title={"IP Rate"}
```
Comment (nomeguy) [inline web/src/RuleEditPage.js]: Label is "IP Rate Limiting"

Comment (nomeguy) [inline web/src/RuleEditPage.js]: i18n

File: web/src/components/IpRateRuleTable.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  }
+
+  restore() {
+    this.updateTable(this.state.defaultRules);
+  }
+
+  renderTable(table) {
+    const columns = [
+      {
+        title: "Name",
```
Comment (nomeguy) [inline web/src/components/IpRateRuleTable.js]: i18n

File: web/src/components/IpRateRuleTable.js [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        key: "name",
+        width: "20%",
+        render: (text, record, index) => (
+          <Input value={record.name} onChange={e => {
+            this.updateField(table, index, "name", e.target.value);
+          }} />
+        ),
+      },
+      {
+        title: "Rate",
```
Comment (nomeguy) [inline web/src/components/IpRateRuleTable.js]: i18n

File: web/src/components/IpRateRuleTable.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        key: "operator",
+        width: "40%",
+        render: (text, record, index) => (
+          <InputNumber style={{"width": "100%"}} value={Number(record.operator)} addonAfter="requests / ip / s" onChange={e => {
+            this.updateField(table, index, "operator", e);
+          }} />
+        ),
+      },
+      {
+        title: "Block Duration",
```
Comment (nomeguy) [inline web/src/components/IpRateRuleTable.js]: i18n

File: web/src/components/UaRuleTable.js [hunk 9]
```diff
@@ -147,7 +147,7 @@ class UaRuleTable extends React.Component {
           <div>
             {this.props.title}&nbsp;&nbsp;&nbsp;&nbsp;
             <Button style={{marginRight: "5px"}} type="primary" size="small" onClick={() => this.addRow(table)}>{"Add"}</Button>
-            <Button style={{marginRight: "5px"}} type="primary" size="small" onClick={() => this.restore()}>{"Restore Build-in User-Agent Rules"}</Button>
+            <Button style={{marginRight: "5px"}} type="primary" size="small" onClick={() => this.restore()}>{"Restore"}</Button>
```
Comment (nomeguy) [inline web/src/components/UaRuleTable.js]: Check all i18n

File: rule/rule.go [hunk 10]
```diff
@@ -36,6 +36,10 @@ func CheckRules(ruleIds []string, r *http.Request) (string, string, error) {
 			ruleObj = &IpRule{}
 		case "WAF":
 			ruleObj = &WafRule{}
+		case "IpRate":
```
Comment (nomeguy) [inline rule/rule.go]: Label is "IP Rate Limiting"

**Review by nomeguy — CHANGES_REQUESTED** [review]

File: controllers/rule.go (hunk 2 above)
Comment (love98ooo) [reply controllers/rule.go:189]: It is not allowed to be zero.

File: controllers/rule.go (hunk 3 above)
Comment (love98ooo) [reply controllers/rule.go]: It has been limited in frontend.

**Review by nomeguy — APPROVED** [review]

**Pull request merged**